doc/ecmp/fine_grained_next_hop_hld.md

```diff
@@ -0,0 +1,306 @@
+    * [1.2 Orchagent requirement](#12-orchagent-requirements)
+    * [1.3 CLI requirements](#13-cli-requirements)
+    * [1.4 Scalability requirements](#14-scalability-requirements)
+    * [1.5 Warm Restart requirements ](#15-warm-restart-requirements)
+  * [2 Modules Design](#2-modules-design)
+    * [2.1 Config DB](#21-config-db)
+    * [2.2 State DB](#22-state-db)
+    * [2.3 CLI](#22-cli)
+    * [2.4 Orchestration Agent](#23-orchestration-agent)
+    * [2.5 SAI](#24-sai)
+  * [3 Flows](#3-flows)
+  * [4 Example configuration](#4-example-configuration)
+  * [5 Warm boot support](#5-warm-boot-support)
+
+###### Revision
+| Rev |     Date    |       Author       | Change Description                |
+|:---:|:-----------:|:------------------:|-----------------------------------|
+| 0.1 | 04/24/2020  |    Anish Narsian   | Initial version                   |
+
+
+# About this Manual
+This document provides the high level design for the Fine Grained ECMP feature implementation in SONiC
+Associated SAI proposal: https://github.com/opencomputeproject/SAI/blob/master/doc/ECMP/Ordered_and_Fine_Grained_ECMP.md
+# Scope
+This document describes the high level design of a staic Fine Grained ECMP feature as implemented in the application layer. 
+- In-scope: Modifying the behavior of ECMP to achieve fine grained handling of ECMP for a static set of next-hops and prefixes as defined in configuration
+- Out of scope: Dynamic ways to enable and use fine grained ECMP
+# Use case
+![](../../images/ecmp/use_case.png)
+
+Firewall or other applications running on loadbalanced VMs which maintain state of flows running through them, such that:
+- There is shared state amongst some set of firewalls so that flows can be recovered, but flow recovery is expensive so we should limit flow redistributions
+- Give that not all firewalls share state, there is a need to redistribute flows only amongst the firewalls which share state
+- An entire firewall set can go down
+
+
+# Definitions/Abbreviation
+###### Table 1: Abbreviations
+| Abbreviation             | Meaning                        |
+|--------------------------|--------------------------------|
+| ECMP                     | Equal Cost MultiPath           |
+| FG                       | Fine Grained                   |
+| NHG                      | Next Hop Group                 |
+
+
+# 1 Requirements Overview
+## 1.1 Functional requirements
+This section describes the SONiC requirements for Fine Grained ECMP next-hop groups 
+
+At a high level the following should be supported:
+
+Phase #1
+- Should be able to configure a Fine Grained ECMP group which defines a static means of redistribution of ECMP upon next-hop modifications based upon the fine grained ECMP SAI proposal
+- Standard route modifications should enable special ECMP behavior for those prefixes which desire Fine Grained ECMP. For all other prefixes the standard ECMP behavior should apply. 

```

Comment:
Not sure if specifying the prefixes to use this feature is a requirement or just the ways to simply the design.  This feature seems to be more tied to the nexthops rather than routes. To manage the static prefixes in configDB is challenging in most networks as routes are normally dynamic learnt, not to mention the potential inconsistency of  routing stack and SONiC config. However, the nexthops are mostly statically configured.  I would think we can make this feature to follow that nature.  I,E, configuration on SONiC are most to enable FG-ECMP for nexthops,  and the FG-ECMP-Group is automatically created if all members are nexthops with FG-ECMP enabled (otherwise, standard ECMP group).  During the route changes, it just uses or creates the ECMP group based on the nexthops the route uses as above.    Does this make any sense?

Reply:
This makes sense and we had considered both options(with and without prefix), from an implementation stand point both would work. We thought that adding the prefix gives more granularity to the configuration: A user has special ecmp behavior for a prefix. At a high level I think that ecmp defined for a prefix so I thought it may make sense to make this config prefix specific. Further it also makes things clearer in terms of expected behavior if we get prefixes with next-hops which are absent from FG_NHG member definitions: ignore those since FG_NHG is a hw feature which makes the final determination of ASIC next-hops based on config pushed by user. Does that make sense? I will also spend some time and rethink if we should remove the prefix.

Reply:
@zhenggen-xu, let me know what you think. We discussed with @lguohan and others as well, along with the reasons highlighted above, the prefix approach may have an advantage for future oriented use cases, for example, special ecmp behavior is set for all prefix announcements bearing a special community tag, given that the community tag is made a part of the db entry. The next-hop approach would then not be compatible with this use case and the prefix based approach would be. Does that make sense?

Reply:
@zhenggen-xu discussed some more on this. We propose to introduce a FG_NHG 'mode' as a future schema change to support this behavior based on any future usecase requirements. The mode specifies whether fgnhg will use a prefix based or next-hop based filtering to determine fine grained behavior:
FG_NHG|{{fg-nhg-group-name}}:
    "bucket_size": {{hash_bucket_size}}
    **"mode":{{prefix/next-hop}}**
If/when the next-hop based implementation is desired, the above schema change may be made and configurable next-hop based filtering enabled. The default mode could remain prefix based in order to allow for backwards compatibility.

Reply:
@anish-n mode will be in TO-do list for future ? Are you planning to update document/schema in the PR ?

Reply:
Added, thanks